pkg/pmem-state/pmme-state.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2019  Intel Corporation.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+package pmemstate
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/klog/glog"
+)
+
+// GetAllFunc callback function used for StateManager.GetAll().
+// This function is called with ID, for each entry found in the state.
+type GetAllFunc func(id string) bool
+
+// StateManager manages the driver persistent state, i.e, volumes information
+type StateManager interface {
+	// Create creates an entry in the state with given id and data
+	Create(id string, data interface{}) error
+	// Delete deletes an entry found with the id from the state
+	Delete(id string) error
+	// Get retrives the entry data into location pointed by dataPtr.
+	Get(id string, dataPtr interface{}) error
+	// GetAll retrieves all entries found in the state, foreach functions is
+	// called with id for every entry found in the state, and entry data is filled in dataPtr.
+	// the caller has to copy the data if needed.
+	GetAll(dataPtr interface{}, foreach GetAllFunc) error
+}
+
+// fileState Persists the state information into a file.
+// This is is supposed to use by Nodes to persists the state.
+type fileState struct {
+	location string
+}
+
+var _ StateManager = &fileState{}
+
+// NewFileState instantiates the file state manager with given directory
+// location. It ensures the provided directory exists.
+// Returns error, if fails to create the direcotry incase of not pre-existing.
+func NewFileState(loc string) (StateManager, error) {
+	if err := ensureLocation(loc); err != nil {
+		return nil, err
+	}
+
+	return &fileState{
+		location: loc,
+	}, nil
+}
+
+// Create saves the volume metadata to file named <id>.json
+func (fs *fileState) Create(id string, data interface{}) error {
+	file := path.Join(fs.location, id+".json")
+	// Create new file for synchronous writes
+	fp, err := os.OpenFile(file, os.O_WRONLY|os.O_SYNC|os.O_CREATE, 0600)
+	if err != nil {
+		return errors.Wrapf(err, "file-state: failed to create metadata storage file %s", file)
+	}
+
+	if err := json.NewEncoder(fp).Encode(data); err != nil {
+		// cleanup file entry before returning error
+		fp.Close() //nolint: errcheck, gosec
+		if e := os.Remove(file); e != nil {
+			glog.Warningf("file-state: fail to remove file %s: %s", file, e.Error())
+		}
+		return errors.Wrap(err, "file-state: failed to encode metadata")
+	}
+
+	return fp.Close()
```

Comment:
What about making these IO operations crash-resilient?

Reply:
Now file writes are synchronous so that no writes to cache. What are other things we should do to make crash-resilient?

Reply:
Here's a good article: https://lwn.net/Articles/667788/

In a nutshell, fsync the file content and fsync the parent directory.

Your crash recovery also needs to be resilient against incomplete files: validate that the file you are reading is complete, remove it if it isn't.

Reply:
> Your crash recovery also needs to be resilient against incomplete files: validate that the file you are reading is complete, remove it if it isn't.

As we are using json decoder, I believe this(incomplete file content) should be handled by json package?

Reply:
Probably you'll get an error because the file isn't valid JSON. But your code then has to have error handling for the broken file.

It's worth writing a unit test for this.

Reply:
Added unit test for this. Done